### ISSUE TITLE
Add support for deleting scopes

### ISSUE THREAD

**Issue by paulmillar** [rucio/rucio#4974]
Motivation
----------

Sometimes a scope that was created is no longer needed.

This might be because the scope was created for testing and isn't needed any more.  It could be because the family of scopes has changed, with data that was located within one scope being moved to different scopes, which leaves one or more scopes redundant.

Currently, this requires manual database intervention, which risks leaving the database in an inconsistent state.

Modification
------------

Add a new admin command `rucio-admin scope delete` to the client and add support for this command on the server side.

An initial version of this command could require that there are no files in the targeted scope.  This is similar to how `rmdir` works.  (an idea due to Andrea Ceccanti).

A subsequent version could support automatic removal of files in that scope.  It might be a good idea to protect against accidentally invoking this variant; for example, by requiring a command-line option.

**Comment by bari12** [comment]
I think this would be fine if the scope is completely empty. 

The issue is if it is not, and even holds deleted files, then we cannot easily delete it due to FK constraints and the conceptually idea that a did is unique for eternity. (Thus if there is a deleted did, it will need the scope as well due to FK)

**Comment by paulmillar** [comment]
If I've understood correctly: Rucio remembers files/datasets/containers that used to exist but have been deleted.  These are the "deleted did"s.  This list of deleted DIDs may prevent Rucio from deleting a scope, since Rucio needs to know the scope for these (deleted) DIDs.

If so, would it make sense for a request to delete a scope to trigger the removal of any deleted DIDs that have the targeted scope?

I think, without this, removing the scope would only work if a scope was created by mistaken and never used to store data (dataset or containers).

**Comment by bari12** [comment]
Exactly, it's due to the foreign key relationships in the database. If there is a data identifier (even deleted) the scope needs to be there.

What you describe would work, but it is a bit of a conceptual question if we want to allow this, until now, we did not. But it would be limited to administrators anyway, so perhaps this is ok.

**Comment by paulmillar** [comment]
If it helps, the concrete motivation came from the ESCAPE project.  In there, we wanted to "clean up" some scopes that were being used for individual/personal testing purposes.  These scopes did not adhere to the ESCAPE naming scheme, so could cause problems.

As I understand it, removing a scope requires manual database intervention and also careful management of the replicas.  The concern is that incorrect database operations could result in dark data if the replicas are not removed correctly.

**Comment by bari12** [comment]
Thank you for raising this. We have reviewed this issue as part of our regular issue review.

We see value in this request, but we do not have the capacity to work on it within the next twelve months.

In line with our [issue management guideline](https://rucio.cern/documentation/developer/issue_management), we are labelling it `wishlist` and closing it. This does not mean it is forgotten: wishlist items are revisited regularly during component reviews and roadmap planning, and this issue can be re-opened and added back to the pipeline as soon as interest and capacity align.
